### DISCUSSION TITLE
Flows, what should I use as a trigger ID

### DISCUSSION THREAD

**Discussion by fa-briz-io** [directus/directus#15814] [Archive]
Hello,

I have a flow where I request coordinates from a webhook/service. The flow is a action and triggers on update and create. It updates the item that triggers it. The thing is, when I create a item the ID I need is $trigger.key, but on update it is array $trigger.keys, so I grab it with $trigger.keys[0]. I could off course create two Flows, but I would like to keep it in one if that is possible.

What is the right way to do this? 

![Screenshot 2022-09-29 at 21 21 04](https://user-images.githubusercontent.com/7761958/193123340-06f538a5-dc35-4eca-91fc-97174096ccf0.jpg)

**Comment by azrikahar** [comment]
This may not necessarily be the right way per se, but I think a potential solution you can try would be to use a Run Script operation which checks for `key` or `keys` from `$trigger`, then always make sure to output an array. So if it's `key` such as `1`, output it as an array `[1]`, else just return `keys` which is already an array.

Now that it is always outputting an array, you can then use the Run Script's output in the Update Data operation's IDs field consistently 👍

**Reply by fa-briz-io** [reply]
Thank you for you input, it is very much appreciated! I decided to use the "condition" block, with the following code.

`{
    "$trigger": {
        "key": {
            "_submitted": "true"
        }
    }
}`

on true I redirect to a "update" block using the ID `$trigger.key` else I assume the ID is in `$trigger.keys[0]`.

This still feels like not such a robust solution, because if the update keys array contains more then one item this does not update anything outside the first one. I think the flow should separate multiple updates like a "for" statement, so that every iteration triggers his own flow.

**Reply by azrikahar** [reply]
> Thank you for you input, it is very much appreciated! I decided to use the "condition" block, with the following code.
> 
> `{ "$trigger": { "key": { "_submitted": "true" } } }`
> 
> on true I redirect to a "update" block using the ID `$trigger.key` else I assume the ID is in `$trigger.keys[0]`.
> 
> This still feels like not such a robust solution, because if the update keys array contains more then one item this does not update anything outside the first one. 

My suggestion of "coercing" the result to an array regardless of key or keys is to make sure there's no need to do `$trigger.keys[0]`, but something like `the_run_script_operation_key` (which contains `[key]`, or `keys`) so it will always be consistent 😁 Unless I'm misunderstanding the logic here of course.

> I think the flow should separate multiple updates like a "for" statement, so that every iteration triggers his own flow.

I believe this should be the case if you use Trigger Flow operation and pass an array to the follow up Flow!

**Reply by fa-briz-io** [reply]
I have still not tried the "run script" block, so I do not really know how it works yet. Also seems there is no info in de documentation about it, probably because it is rather new. I'll try and figure it out later on, as this seems like a good suggestion!

> I believe this should be the case if you use Trigger Flow operation and pass an array to the follow up 

I did not think of this, thanks a lot!

**Reply by vtoomas** [reply]
I did it as suggested @azrikahar using 2 flows and will try describing in bit more detail. 

**Flow 1 - Process each key** - This Flow is run for each key (like a for loop body). 
Trigger is Another Flow (Response body doesn't matter unless you want to return data from other flow I believe). 
In the next Operation you can do whatever to the current key by referring to it with just {{$trigger}}

**Flow 2 - Trigger for created and update events** - This Flow is calling Flow 1 with each key. 
Trigger on created and update events. Followed by Run Script block which always returns array of keys, either there is key or keys from Event trigger. Code here (thanks @SimonHausdorf , modified your code from below): 
```
module.exports = async function(data) {
	if(data['$trigger']['keys'] == undefined) {
        return [data['$trigger']['key']];
    } else {
        return data['$trigger']['keys'];
    }
}
```
Followed by Trigger Flow block with payload {{$last}} (make sure you have Toggle Raw active) and Flow 1 selected.

**Comment by shipsides-sparknel** [comment]
this is definitely something i consider a bit of a gap as well.

aside from running a script to get the non-null, you could make an update data action for {{ $trigger.key }}, duplicate the action, and set the fail branch as the duplicate for {{ $trigger.keys[0] }}

**Reply by fa-briz-io** [reply]
This is also a great suggestion, thanks for you input!

**Comment by SimonVillage** [comment]
I came up with the following solution.

Add a script like so and add a key as `get_key`
```javascript
module.exports = async function(data) {
    if(data['$trigger']['keys'] == undefined) {
        return {id: data['$trigger']['key']};
    } else {
        return {id: data['$trigger']['keys'][0]};
    }
}
```

now you can access your id with `{{get_key.id}}`
